### ISSUE TITLE
event table in osqueryd: inconsistent event query results

### ISSUE THREAD

**Issue by jfdive** [facebook/osquery#2467]
when doing tests on events generation with the syslog table, i faced a problem i could not solve. After chatting around on #slack it seems to be agreed there is a problem. 

* problem description:
-> inconsistent event count when doing select in event table for a determined set of events generated manually.
-> events are missed or duplicated.

* test setup / reproduction steps
-> ubuntu 16.04
-> git osqueryd or osquery.io provided package
-> osqueryd.conf:
    "options":
    {   
        "config_plugin": "filesystem",
        "logger_plugin": "filesystem",
        "logger_path": "/var/log/osquery",
        "pidfile": "/var/run/osquery.pidfile",
        "events_expiry": "3600",
        "database_path": "/var/osquery/osquery.db",
        "verbose": "true",
        "worker_threads": "2",
        "enable_monitor": "true",
        "schedule_splay_percent": "0"
    },
        "syslog_sudo_cmd":
        {   
            "query": "SELECT time,facility,tag,message FROM syslog WHERE tag = 'sudo' AND message like '%COMMAND=%';",
            "interval": 10,
            "removed" : false
        }
-> test command: for i in {1..50}; do sudo echo zwass_$i; sleep 1; done
-> expected to see 50 logs in results.log
-> sudo grep zwass /var/log/osquery/osqueryd.results.log  | cut -d ";" -f 4 | grep echo | cut -d "," -f 1 | cut -d " " -f 3 | cut -d '"' -f 1
...
zwass_26
zwass_27
zwass_28
zwass_27
zwass_28
zwass_29
zwass_30
zwass_31
...
-> sudo grep zwass /var/log/osquery/osqueryd.results.log  | cut -d ";" -f 4 | grep echo | cut -d "," -f 1 | cut -d " " -f 3 | cut -d '"' -f 1 | wc -l 
58

-> event table seems to be fine: stopping osqueryd and querying using osqueryi --database_path= shows the correct event code.

* where is the problem:
-> potentially in the bug would be with the differential query logic.

**Comment by theopolis** [comment]
Ok, I can reproduce this after following the `rsyslog` configuration: http://osquery.readthedocs.io/en/stable/deployment/syslog/

Using the configuration:
```json
{
  "options": {
    "config_plugin": "filesystem",
    "logger_plugin": "filesystem",
    "events_expiry": "3600",
    "verbose": "true",
    "enable_monitor": "true"
  },
  "schedule": {
    "syslog_sudo_cmd": {
      "query": "SELECT time,facility,tag,message FROM syslog WHERE tag = 'sudo' AND message like '%COMMAND=%';",
      "interval": 10,
      "removed": false
    }
  }
}
```

And invoking a debug/development `osqueryd`:
```
sudo ./build/linux/osquery/osqueryd \
  --verbose --disable_events=false --ephemeral \
  --database_path /tmp/osquery/osquery.db --logger_path /tmp/osquery \
  --allow_unsafe --config_path ./build/syslog.conf
```

In case anyone is interested in following along.

**Comment by theopolis** [comment]
There are several issues here. The first is the conversion of syslog time to an osquery-standard epoch time. Depending on the syslog setup, osquery may drop events: https://github.com/facebook/osquery/blob/master/osquery/events/linux/syslog.cpp#L206

To fix this we should move the "syslog view" of time/seconds into a `datetime` column and reserve `time` for the UNIX epoch, which does not account for DST or TZ configuration.

**Comment by theopolis** [comment]
To fix this more generally I am going to remove all event subscriber ability to dictate time to the database. If specific subscribers, like `syslog` want to emit a time-other-than-epoch-in-seconds they can pick and emit to another column.

**Comment by theopolis** [comment]
The last issue comes from the "optimization" concept, which was invented while all event-data was also applied to a set-difference similarly to non-event tables.

Optimization occurs by default and introduces a sliding window for storing and retrieving events in response to the configured schedule intervals. The "last time you queried" is saved `-1` second such that the next time you query the index retrieval is from `t - interval - 1`.

The issue is `-1` without a set difference means any events emitted last query within `(t - interval - 1, t - interval)` will be duplicated. :/

**Comment by theopolis** [comment]
See #2508 for the fix, and I've opened #2513 for another problem I noticed.
